# More work on the RNAseq data with BenE and BenF

Since our last meeting Xue has trimmed the sequence data from X. laevis and X. tropicalis using trimmomatic.  We now want to do the following:
- [x] Set up a 'vanilla' trinity assembly for each transcriptome
- [x] Download XL and XT genomes using wget
- [x] Make a blast database out of each genome.  This will be used to assess quanity of the transcriptome assemblies from trinity.


# Summary of chat with BenF, BenE, and Xue

## Issues related to the technical accuracy of the assembly program
* Develop a metric for assembly fidelity based on BLAST (or BLAT) results for each transcriptome query.  This might, for example, include an assessment of whether or not all exons map to one chromosome, and whether they are in the proper order.
* Assess the impact of rate of evolution on the assembly fidelity
* Directions for the future: assess performance of other software

## Issues related to the biology of our organisms (i.e. diploid versus tetraploid
* For 'real' transcripts, do we see more alternate splicing in the tetraploid than the diploid?

# Trinity assembly

First in this directory (`/home/xue/transcriptome_data/BJE3909_tropicalis_trimmed_data`), we combined the trimmed forward and reverse reads like this:
```
zcat BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_001_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_002_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_003_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_004_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_005_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_006_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_007_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_001_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_002_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_003_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_004_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_005_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_006_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_007_paired.fastq.gz > BJE3909_paired_R1.fq
```
```
zcat BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_001_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_002_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_003_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_004_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_005_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_006_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_007_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_001_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_002_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_003_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_004_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_005_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_006_paired.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_007_paired.fastq.gz > BJE3909_paired_R2.fq
```

```
zcat BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_001_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_002_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_003_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_004_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_005_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_006_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R1_007_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_001_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_002_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_003_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_004_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_005_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_006_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R1_007_single.fastq.gz  BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_001_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_002_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_003_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_004_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_005_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_006_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L004_R2_007_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_001_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_002_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_003_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_004_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_005_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_006_single.fastq.gz BenEvansBJE3909cDNA_Library_GTGAAA_L005_R2_007_single.fastq.gz > BJE3909_single_R1_and_R2.fq
```

Now each of these files needs to be compressed like this:
```
gzip BJE3909_paired_R1.fq
gzip BJE3909_paired_R2.fq
gzip BJE3909_single_R1_and_R2.fq
```

The path to the compressed files are:
```
/4/xue/transcriptome_data/BJE4168_laevis_trimmed_data/BJE4168_paired_R1.fq.gz
/4/xue/transcriptome_data/BJE4168_laevis_trimmed_data/BJE4168_paired_R2.fq.gz
/4/xue/transcriptome_data/BJE4168_laevis_trimmed_data/BJE4168_paired_R1.fq.gz
/4/xue/transcriptome_data/BJE4168_laevis_trimmed_data/BJE4168_paired_R2.fq.gz
```

# Fastqc

You can make a html file with QC information like this:

```
fastqc BJE3909_single_R1_and_R2.fq.gz
```

and then download the file to inspect with a browser like this:

```
scp xue@info.mcmaster.ca:/home/xue/transcriptome_data/BJE3909_tropicalis_trimmed_data/BJE3909_single_R1_and_R2_fastqc.html .
```


# Starting up a trinity assembly

```
/home/xue/trinityrnaseq-2.2.0/Trinity --seqType fq --left /home/xue/transcriptome_data/BJE3909_tropicalis_trimmed_data/BJE3909_paired_R1.fq.gz --right /home/xue/transcriptome_data/BJE3909_tropicalis_trimmed_data/BJE3909_paired_R2.fq.gz --single /home/xue/transcriptome_data/BJE3909_tropicalis_trimmed_data/BJE3909_single_R1_and_R2.fq.gz --CPU 6 --max_memory 20G
```
2 additional function added: add cleanup flag to skip all the intermediate files and keep the fastq file only; and the output directory is named:
```
/home/xue/trinityrnaseq-2.2.0/Trinity --seqType fq --left /home/xue/transcriptome_data/BJE4168_laevis_trimmed_data/BJE4168_paired_R1.fq.gz --right /home/xue/transcriptome_data/BJE4168_laevis_trimmed_data/BJE4168_paired_R2.fq.gz --single /home/xue/transcriptome_data/BJE4168_laevis_trimmed_data/BJE4168_single_R1_and_R2.fq.gz --CPU 6 --max_memory 20G --full_cleanup --output /home/xue/transcriptome_data/laevis_trinity_out 
```
# Downloading genome data from Xenbase
Genome data for X.tropicalis (Version 9) and X.laevis(version 9.1) were downloaded from Xenbase and is store in (`/home/xue/genome_data`)
```
wget ftp://ftp.xenbase.org/pub/Genomics/JGI/Xentr9.0/Xtropicalis.v9.repeatMasked.fa.gz
```
The name of the file cannot be copy&paste on Xenbase, a quick way to get the file is to use * to select the name of the file with specific word
```
wget ftp://ftp.xenbase.org/pub/Genomics/JGI/Xenla9.1/*repeat*
```
#Making a blast database using downloaded genome
The genome data file is in gz and makeblastdb only run with fasta file, so the genome data was unzip using:
```
gunzip Xtropicalis.v9.repeatMasked.fa.gz
```
the blast database is made using:
```
makeblastdb -in Xla.v91.repeatMasked.fa -dbtype nucl -out db_Xlaevis_v91
```
```
makeblastdb -in Xtropicalis.v9.repeatMasked.fa -dbtype nucl -out db_Xtropicalis_v9
```
Output information after blast databases shown the # of sequences added: 402501 sequences for X.laevis; 6823 sequences for X.tropicalis.

#Running local blastn
The assembled transcriptome data are blasted against its respective local blast database. Blastn is performed with: 
```
blastn -task blastn -db /home/xue/genome_data/db_Xtropicalis_v9 -outfmt 6 -evalue 0.00005 -query /home/xue/transcriptome_data/Assembled_BJE3909_Tropicalis_Trinity.fasta -out /home/xue/transcriptome_data/blastn_BJE3909_output/blastn_BJE3909
```
#Creating testfile
```
head -n 500000 blastn_BJE3909 > testfile500k
```
creat testfile from zipped file:
```
head -n 500000 <(blastn_BJE3909.gz) > testfile500k
```
