# Testing out Xue's new script!

First, we need to make sure that the script is executable like this:

```
chmod 755 BlastnOutputAnalysis.pl
```

To execute the script locally (not on server):
```
./BlastnOutputAnalysis.pl
```
To put perl script on the server so that I can run it on the server:
```
scp BlastnOutputAnanlysis_Oct24_v3_clean.pl  xue@info.mcmaster.ca:/home/xue/transcriptome_data/blastn_BJE3909_output
```
#To run the script with a zipped (gz) file:
*run with default parameter
```
perl BlastnOutputAnalysis_3909_Feb20.pl <(zcat blastn_BJE3909_run2.gz) > 3909_run2_ErrorDetial;echo "3909 run2 started at 10:25 on Feb 28 2017"|mail sarahsongxy@gmail.com
```
*Defined name of output files and lenght limit of overlapping part by the user:
```
perl BlastnOutputAnalysis_3909_Feb20.pl <(zcat blastn_BJE3909_run2.gz) -OLL=10 -out1=3909_run2_ErrorFreelist_OLL5 - out2=3909_run2_Detial_OLL5 > 3909_run2_ErrorDetial;echo "3909 run2 started at 10:25 on Feb 28 2017"|mail sarahsongxy@gmail.com

```

#Meeting - Mar 13 2017
* check if L/S error is due to missing infor on genome database or transcript assembly error
[] compare identity%, defined as diverge if the difference is >7%
[] if <7%, probably is an assembly error
[] if >7%, could be due to missing information in teh genome database
[] to confirm the reason, extract the fragment which is with lower identity% and blastn it to the genome. If it only mapped to L (while the rest of the fragment in the transcript is S), it is due to missing info 
* ask Caroline about trans-abyss


