# Assessment Metric Overview Flowchart

# Main 
1. the input file for the metric is the Blastn output file. The below is the format for blastn output 
   - blast format 6: qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore
   - qseqid is the name of the transcripts being aligned against the genome
   - sseqid is the name of chromosome/scafold that the transcript aligned to
   - qstart and qend of transcripts(qseqid) was aligned to sstart and send of the chromosome/scaffold (sseqid) 
2. A while loop was used to read the input file line by line. Each line is a hits for a qseqid. 
   - a hash named %transcript_hash was used to stored the input line and the below is the structure of the hash:
   - hash structure: qseqid, fragment number, ties number, array for stats storage(last level)
    - last level array stores: 0=sseqid; 1=qsatart; 2=qend; 3=sstard; 4= send; 5=evalue; 6=bitsore; 7=forward(1)/reverse(-1); 8=length;
   - ex: $transcript_hash{qseqid}{fragment id}{ties id}[0] = chr04;
3. New input line will be compared to existing fragments in the hash. The e-value of the new line will be used to determine if the line will be store in the hash or not. We selected non-overlapping fragment (5bp of overlapping is allowed) with the lowest e-value. 
    - once a fragment is selected, it will be added to the hash with a new fragment number by calling a subfunction called addfragment.
   - if the selected fragment is tie, it will be added to the hash with a new tie number by calling a subfunction called addties. 
4. Once all the lines/hits for a transcript has been selected and evaluated, subfunction checkerror_tropicalis or sortties_tropicalis if the transcript contain ties will be called to evaluate whether the transcript is assemblied correctly. The assessment result(error free or error types) will be stored in a hash called %error-hash. There were 4 type of error messages:
   - "no error": the transcript was error free. It had the same chromosome/scaffold ID, 
   - "Contain scaffold“: the transcript contains fragments that mapped to chromosome and scaffolds. The transcript would not be evaluate for orientation error or order error since the genome coordinate of fragments that was mapped to scaffolds would definately not in order with the genome coordinate of fragments that were mapped to the chromosome. 
   - "Chromosome Error": fragments were mapped to different chromosomes.
   - "Orientation Error": fragments were assembled in different orientation.
   - "Order Error": fragments were not in ascending order(if forward direction)  or descending order (if reverse direction)
 5. Once the evaluation result has returned back from the subfunction, %transcript_hash will be cleared and start loading hits for a new qsedid. 
 6. Once assessment of all the transcripts has been done, assessment result in %error-hash will be summarized. The script will print the list of qseqid(transcript names) that is flagged as error free and the previously mentioned summary will be print at the end of the list.  
 
 # Subfunction subfunction checkerror_tropicalis 
 Please keep in mind that this subfunction assess transcripts that contains fragments with no ties. 
 1. The fragments will be restored in ascendign order based on qstart position. The error message associated with each transcript will be stored in a hash called %error-hash. Its structure is like $error-hash{transcript id} = "error message". Default error message is "no error".
    - reason for doing this: During the process of fragment selection (Main step 3), some fragment might be deleted. Hence, the fragment numbers might not be consecutive. Additionally, when the fragments were selected, they might not selected in the same order as they are in the transcript.  
2. Check for chromosome error. We used the chromosome ID(chrID) of each fragment to answer this question. 
- 2.1 if it X.tropicalis transcriptome:
   - In X.tropicalis genome, scaffold 1-10 are located in chromosome 1-10 respectivefully. Hence, 4 situations can happen when assessing the transcript for chromosome error. 
   - case 0(default): chrIDs were the same. It would be flagged as "no error".
   - case 1: chrIDs were scaffold >10 and same chromosome (ex.scaffold 175 and chromosome 2), it would be flagged with "Contain scaffold".
   - case 2:  chrIDs were scaffold 1-10 and chromosome 1-10 but they had the same number (ex, scaffold 9 and chromosome 9). This transcript was flagged with "Contain scaffold". 
   - case 3:chrIDs were scaffold 1-10 and chromosome 1-10, and the number in the chrID does not match (ex. scaffold 9 and chromosome 1).  it would be flagged with "Problem: different chromosome";
    
3. Check for Orientation error. We used the direction that was pre-determined when the fragments were added to the hash. Forward is 1 and reverse is -1. 
   - check all fragments, if all fragments was mapped to the genome in same directions, it would be flagged as "no error". 
   - if not, it will be flagged as "Problem:different orientation".

4. check if fragments are assemblied in the proper ascending order or descending order order. We are checking it in the fragment order (1->10).
   - if the direction is 1 (forward), the genome coordinate (sstart and send) need to be a larger number compared to its previous fragments.
   - if the direction is -1 (reverse), the genome coordinate (sstart and send) need to be a larger number compared to its previous fragments.

5. If the transcript passed all three check, it will be associated with the default error message "no error" in %error_hash. 
 
 
 # Subfunction sortties_tropicalis
 2. 
 
