# Assessment Metric Overview Flowchart

# Main 
1. the input file for the metric is the Blastn output file. The below is the format for blastn output 
  - blast format 6: qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore
  - qseqid is the name of the transcripts being aligned against the genome
  - sseqid is the name of chromosome/scafold that the transcript aligned to
  - qstart and qend of transcripts(qseqid) was aligned to sstart and send of the chromosome/scaffold (sseqid) 
2. A while loop was used to read the input file line by line. Each line is a hits for a qseqid. 
  - a hash named %transcript_hash was used to stored the input line and the below is the structure of the hash:
  - hash structure: qseqid, fragment number, ties number, array for stats storage(last level)
  - last level array order: 0=sseqid; 1=qsatart; 2=qend; 3=sstard; 4= send; 5=evalue; 6=bitsore; 7=forward(1)/reverse(-1); 8=length;
  - ex: $transcript_hash{qseqid}{fragment id}{ties id}[0] = chr04;
3. New input line will be compared to existing fragments in the hash. The e-value of the new line will determine if the line will be store in the hash or not.
4. Once all the lines/hits for a transcript has been evaluate and selected, subfunction checkerror_tropicalis or sortties_tropicalis if the transcript contain ties will be called to evaluate whether the transcript is assemblied correctly. 
 5. Once the evaluation result has returned back from the subfunction, %transcript_hash will be cleared and start loading hits for a new qsedid. 
 
 # Subfunction subfunction checkerror_tropicalis 
 
 # Subfunction sortties_tropicalis
 
