#Identify Splice Variants
#Conversation Summary
Summary from Jan 9 and 30 2017 discussion between Dr Ben Evans and Xue Song regarding identification of splice variants.
- goal: identify total exons, max exons per transcript, min exons per transcript, location of variants(ends or middle)
- method: 
  - firstly, run Blast with set of assembled transcripts flagged with "no error" or "contain scaffold" against itself. 
  - Then used the Blast result and coordinate information from previous genomic blast to achieve goal
- critiria
  - transcripts that are just repeat region of each other does not count as splicing vairants. 
  - transcripts that identified as splicing variants should have genomic coordinate close to each other
  - transcripts that identified as splicing variants have to hae two exons mapped to each other or have one large mapped exon that is greater than 200bp(changable variable). 
# Extract Error-free transcripts 
Assessment metric will output a list of name of Error-Free transcripts named and stored in the following file:
```
/home/xue/transcriptome_data/blastn_BJE4168_output/4168_ErrorFreeList 
```
A customized script is written to extract the sequence of error-free transcripts from the trinity output file. 
The script is stored in:
```
/home/xue/transcriptome_data/IdentifySplicingVariants/ExtractTranscriptome.pl*
```
The command for the extraction is:
```
perl ExtractTranscriptome.pl /home/xue/transcriptome_data/blastn_BJE4168_output/4168_ErrorFreeList Assembled_BJE4168_Laevis_Trinity.fasta
```
# Building local database with extracted sequence of error-free transcripts
The local database was built with the following command:
For tropicali:
```

```
For laevis:
```
```

# Blastn with Error-free transcripts
Command for running blastn with db and error-free transcripts
```
blastn -task blastn -db db_4168_NoErrorTranscripts -outfmt 6 -evalue 0.00005 -query ExtractedAssembled_BJE4168_Laevis_Trinity.fasta  -out blastout_4168; echo "blastn of 4168 extrated transcriptome to itself at 14:25  Feb 27 2017"|mail sarahsongxy@gmail.com

```
# Run the script for identifying Alternative Splicing Variants:
For tropicali (3909):
```
perl IdentifySplicingVariants.pl /home/xue/transcriptome_data/blastn_BJE3909_output/3909_Detail_OLL5 blastout_3909 >test_print_3909

```
For Laevis(4168):
```
perl IdentifySplicingVariants.pl /home/xue/transcriptome_data/blastn_BJE4168_output/4168_Detail blastout_4168 > test_print_4168
```
 
