#Identify Splice Variants
#Conversation Summary
Summary from Jan 9 and 30 2017 discussion between Dr Ben Evans and Xue Song regarding identification of splice variants.
- goal: identify total exons, max exons per transcript, min exons per transcript, location of variants(ends or middle)
- method: 
  - firstly, run Blast with set of assembled transcripts flagged with "no error" or "contain scaffold" against itself. 
  - Then used the Blast result and coordinate information from previous genomic blast to achieve goal
- critiria
  - transcripts that are just repeat region of each other does not count as splicing vairants. 
  - transcripts that identified as splicing variants should have genomic coordinate close to each other
  - transcripts that identified as splicing variants have to hae two exons mapped to each other or have one large mapped exon that is greater than 200bp(changable variable). 
  
  
#Blastn with Error-free transcripts
Command for running blastn with db and error-free transcripts
```
blastn -task blastn -db db_4168_NoErrorTranscripts -outfmt 6 -evalue 0.00005 -query ExtractedAssembled_BJE4168_Laevis_Trinity.fasta  -out blastout_4168; echo "blastn of 4168 extrated transcriptome to itself at 14:25  Feb 27 2017"|mail sarahsongxy@gmail.com

```
  
 
