# Running Bowtie 2
With paired-end data for X.laevis:
```
bowtie2 --local --no-unal -x Assembled_BJE4168_Laevis_Trinity.fasta -1 BJE4168_paired_R1.fq.gz -2 BJE4168_paired_R2.fq.gz \ | samtools view -Sb - | samtools sort -no - - >bowtie2.4168.bam; echo "quality assesment of 4168 is done, started 8:46" |mail sarahsongxy@gmail.com
```
Below is the result for X.laevis: 
```
SAM header is present: 123547 sequences.
54041010 reads; of these:
  54041010 (100.00%) were paired; of these:
    1539831 (2.85%) aligned concordantly 0 times
    14316327 (26.49%) aligned concordantly exactly 1 time
    38184852 (70.66%) aligned concordantly >1 times
    ----
    1539831 pairs aligned concordantly 0 times; of these:
      127200 (8.26%) aligned discordantly 1 time
    ----
    1412631 pairs aligned 0 times concordantly or discordantly; of these:
      2825262 mates make up the pairs; of these:
        1121524 (39.70%) aligned 0 times
        337141 (11.93%) aligned exactly 1 time
        1366597 (48.37%) aligned >1 times
98.96% overall alignment rate
```
```
[106,900,000]  lines read


#read_type      count   pct
proper_pairs    52751321        98.48
improper_pairs  641463  1.20
left_only       123480  0.23
right_only      51448   0.10

Total aligned rnaseq fragments: 53567712

```

Below are the result for X.tropicali:
```
  47888199 (100.00%) were paired; of these:
    1435402 (3.00%) aligned concordantly 0 times
    20764864 (43.36%) aligned concordantly exactly 1 time
    25687933 (53.64%) aligned concordantly >1 times
    ----
    1435402 pairs aligned concordantly 0 times; of these:
      209906 (14.62%) aligned discordantly 1 time
    ----
    1225496 pairs aligned 0 times concordantly or discordantly; of these:
      2450992 mates make up the pairs; of these:
        934643 (38.13%) aligned 0 times
        322604 (13.16%) aligned exactly 1 time
        1193745 (48.70%) aligned >1 times
99.02% overall alignment rate
```

```

[94,800,000]  lines read


#read_type      count   pct
proper_pairs    46772423        98.45
improper_pairs  559882  1.18
left_only       131968  0.28
right_only      45177   0.10

Total aligned rnaseq fragments: 47509450

```
Trinity git wiki page stated taht "A typical Trinity transcriptome assembly will have the vast majority of all reads mapping back to the assembly, and ~70-80% of the mapped fragments found mapped as proper pairs.". Assemblies for X.tropicalis and X.laevis both had >98% of the mapped fragment found mapped as proper pairs. This indicate that those Trinity assemblies had very high quality. 



#Trinity Transcriptome Contig Nx Statistics
N50 value: at least half of all assembled bases are in transcript contigs of at least the N50 length value.
Below commend was used to run script that is in the Trinity toolkit and is able compute these stats. 
```
perl /home/xue/trinityrnaseq-2.2.0/util/TrinityStats.pl Assembled_BJE4168_Laevis_Trinity.fasta >Stats4168
```
Below are the result for X.laevis:
```
################################
## Counts of transcripts, etc.
################################
Total trinity 'genes':  103447
Total trinity transcripts:      123547
Percent GC: 42.15

########################################
Stats based on ALL transcript contigs:
########################################

        Contig N10: 3337
        Contig N20: 2601
        Contig N30: 2077
        Contig N40: 1640
        Contig N50: 1253

        Median contig length: 423
        Average contig: 755.40
        Total assembled bases: 93327586
#####################################################
## Stats based on ONLY LONGEST ISOFORM per 'GENE':
#####################################################

        Contig N10: 3072
        Contig N20: 2261
        Contig N30: 1687
        Contig N40: 1240
        Contig N50: 910

        Median contig length: 379
        Average contig: 636.96
        Total assembled bases: 65891216
```
Below are the result for X.tropicalis:
```
################################
## Counts of transcripts, etc.
################################
Total trinity 'genes':  80964
Total trinity transcripts:      94772
Percent GC: 42.96

########################################
Stats based on ALL transcript contigs:
########################################

        Contig N10: 5170
        Contig N20: 3938
        Contig N30: 3103
        Contig N40: 2487
        Contig N50: 1939

        Median contig length: 436
        Average contig: 953.63
        Total assembled bases: 90377469
#####################################################
## Stats based on ONLY LONGEST ISOFORM per 'GENE':
#####################################################

        Contig N10: 4773
        Contig N20: 3516
        Contig N30: 2702
        Contig N40: 2096
        Contig N50: 1580

        Median contig length: 394
        Average contig: 816.89
        Total assembled bases: 66138689
```

